### ISSUE TITLE
Swipe action is executed before swipe exposes action buttons

### ISSUE THREAD

**Issue by kenargo** [mikepenz/FastAdapter#934]
- Briefly describe the issue

I have a list, based on SwipeDrawerListActivity sample and things work well except for an oddity I'm seeing which the sample does not implement.

In my use case the item is selectable.  Things (mostly) work, I can swipe to expose the underlying swipe actions, I get notified when the swipe actions are touched, etc.  

Since my items are selectable there needs to be selection management.  I found a few issues:

* If I use the sample code (`constraintViewJobSiteAssignmentWithIDAdapter.background = FastAdapterUIUtils.getSelectableBackground(ctx, Color.RED, true)`) then I find that the underlying swipe actions come through  and are displayed at the same time as the main list item content.  I got around this by coding my own 'selector' though it isn't perfect.

* If I set the main list content as 'clickable', as the sample shows, then the item is never selected, if I don't set clickable, then selection works (except see next item)

* If I touch the main list item, in the area where the underlying swipe action is contained the swipe action is executed, even though the swipe panel isn't shown.  

I have code in place that blocks the swipe actions from working for non-selected items but I haven't found a way to block swipe actions from executing on selected items and when the swipe area isn't opened.  It's almost like a touch action in the call path if returning false when it should return true and touch actions are falling to the swipe panel and not the main panel (but I'm guessing)

Everything is the latest, your libraries, Android Studio, etc.

I've not found any related issues in the issues tree so I'm resorting to opening an issue.

Oh, yes, it repros in sample app.  Just add this to the SwipteDrawerListActivity:

```
        val selectExtension = fastItemDrawerAdapter.getSelectExtension()
        selectExtension.isSelectable = true
```

Notice touching an item doesn't send click sound.  But if you remove the clickable and focusable from the LinearLayout, then it can select and you will see that touching the area in the list item that hides the swipe actions will activate.

and adding this to the SwipeableDrawerItem shows the issue with overlapping text:

```
val ctx = holder.itemContent.context
holder.itemContent.background = FastAdapterUIUtils.getSelectableBackground(ctx, Color.RED, true)

```

**Comment by kenargo** [comment]
Mike, what does "help wanted" mean?  Are you waiting on something from me?

**Comment by mikepenz** [comment]
@kenargo it mainly means that I am very busy at the very moment, and I wasn't able to have a look over the weekend. hoping that possibly somebody else will chime in.

**Comment by kenargo** [comment]
OK, I'll take a look; I'd also like to dynamically block swipe per item at runtime, maybe I can get this fixed and that added at the same time.  If you get to it let me know, just so we are not duplicating efforts.

**Comment by kenargo** [comment]
Almost; ugly but it works:

I added callbacks for each action and an extra one to talk to the activity like this:

```
    var onClickToSelect: (JobSiteAssignmentWithIDAdapter) -> Unit = {}
    var onClickEdit: (AssignmentRecordWithID) -> Unit = {}
```

The, in the override for bindView I did this:

`editActionRunnable = Runnable { if (item.isSelected) { item.onClickEdit(item.assignmentWithIDAdapter!!) } else {  item.onClickToSelect(item)} }`

In addition, to block clicks from being directed to the actions before the panel is expanded I added `android:clickable="true"` to the main item layout (the one that covers the swipe panel) and then redirected its onClick the same way:

```
mainlayoutView.setOnClickListener {
       item.onClickToSelect(item)
}

```

It means that all selection UI has been delegated to the owner (activity or fragment) but otherwise it's working.

It would be nice to have some item call that asks if an adapter item can open (aka swipe) at runtime, I'd like to stop swipe unless selected) but that's less important right now.  If you have any ideas on this latter question, I'm all ears.

Thanks,
Ken

**Comment by mikepenz** [comment]
@kenargo glad you were able to solve it. Ultimately these things go all back to the APIs google provide :/ So it's less about the `FastAdapter` or the extension we provide to try to make the 99% usecases simple. 

Look into the `ItemTouchHelper`, .. and other associated classes from Android which are used to provide this functionality

**Comment by mikepenz** [comment]
Closing as it was solved, adding the good to know label for others to explore
